bindings.c

```diff
@@ -4092,6 +4292,363 @@ static int proc_swaps_read(char *buf, size_t size, off_t offset,
 	return rv;
 }
 
+/*
+ * Find the process pid from cgroup path.
+ * eg:from /sys/fs/cgroup/cpu/docker/containerid/cgroup.procs to find the process pid.
+ * @pid_buf : put pid to pid_buf.
+ * @dpath : the path of cgroup. eg: /docker/containerid or /docker/containerid/child-cgroup ...
+ * @depth : the depth of cgroup in container.
+ * @sum : return the number of pid.
+ * @cfd : the file descriptor of the mounted cgroup. eg: /sys/fs/cgroup/cpu
+ */
+static int calc_pid(char ***pid_buf, char *dpath, int depth, int sum, int cfd)
+{
+	DIR *dir;
+	int fd;
+	struct dirent *file;
+	FILE *f = NULL;
+	size_t linelen = 0;
+	char *line = NULL;
+	int pd;
+	char *path_dir, *path;
+	char **pid;
+
+	/* path = dpath + "/cgroup.procs" + /0 */
+	path = (char *)malloc(strlen(dpath) + 20);
+	if (!path) {
+		lxcfs_error("Failed to malloc %s in calc_pid!\n", dpath);
+		return sum;
+	}
+
+	strcpy(path, dpath);
+	fd = openat(cfd, path, O_RDONLY);
+	if (fd < 0)
+		goto out;
+
+	dir = fdopendir(fd);
+	if (dir == NULL)
+		goto out;
+
+	while (((file = readdir(dir)) != NULL) && depth > 0) {
+		if (strncmp(file->d_name, ".", 1) == 0)
+			continue;
+		if (file->d_type == DT_DIR) {
+			/* path + '/' + d_name +/0 */
+			path_dir = (char *)malloc(strlen(path) + 2 + sizeof(file->d_name));
+			if (!path_dir) {
+				lxcfs_error("Failed to malloc %s in calc_pid!\n", path);
+				continue;
+			}
+			strcpy(path_dir, path);
+			strcat(path_dir, "/");
+			strcat(path_dir, file->d_name);
+			pd = depth - 1;
+			sum = calc_pid(pid_buf, path_dir, pd, sum, cfd);
+			free(path_dir);
+		}
+	}
+	closedir(dir);
+
+	/* realloc() failed */
+	if (sum == -1)
+		goto out;
+
+	strcat(path, "/cgroup.procs");
+	fd = openat(cfd, path, O_RDONLY);
+	if (fd < 0)
+		goto out;
+
+	f = fdopen(fd, "r");
+	if (!f)
+		goto out;
```

Comment:
You're leaking `fd` if `fdopen()` fails.

Reply:
OK, I will rewrite them.